File: Python/DecodeWays.py
def decodeWays(s):
    dp = {len(s) : 1}

    def dfs(i):
        if i in dp:
            return dp[i]

        if s[i] == '0':
            return 0

        dp[i] = dfs(i+1)

        if i+1<len(s) and (s[i] == '1' or (s[i] == '2' and s[i+1] in "0123456")):
            dp[i]+=dfs(i+2)

        return dp[i]

    return dfs(0)

File: Javascript/DecodeWays.js
var decodeWays = function(s){
    const dp = new Array(s.length+1).fill(1)
    
    for (let i = s.length-1;i>=0;i--){
        if (s[i] === '0'){
            dp[i] = 0
        }
        else{
            dp[i] = dp[i+1]
        }

        if (i+1<s.length && (s[i] === '1' || (s[i] === '2' && "0123456".includes(s[i+1])))){
            dp[i] += dp[i+2]            
        }
    }

    return dp[0]
}